### ISSUE TITLE
Feature requests

### ISSUE THREAD

**Issue by asdf23** [scanmem/scanmem#231]
I'm interested in helping to develop if you need help. 
I'd like to add the following features:

- An ability to tag memory regions (I'd like to add a column to the list command, and then in the set command an ability to specify a specific set of memory region(s) with said "tags")
- An ability to store and retrieve a specific snapshot. I want to take a snapshot and know it is saved as "snapshot0" unless I specific a different name, then load said snapshot at a later point in time.
- An ability to restrict the start/offset of the snapshot/scan taken
- An ability to execute a series of commands concurrently

I can explain it better by example:

```
$ scanmem
0> 6
info: we currently have 47737 matches.
47737> 6
47037> =
8> list
[ 0] 3035fdd48f29,  0 +   54bf29,   exe, 6, [I8 ]
[ 1] 3035fdd48f31,  0 +   54bf31,   exe, 6, [I8 ]
[ 2] 3035fdd48f69,  0 +   54bf69,   exe, 6, [I8 ]
[ 3] 3035fdd48f71,  0 +   54bf71,   exe, 6, [I8 ]
[ 4] 3035fdd48fa9,  0 +   54bfa9,   exe, 6, [I8 ]
[ 5] 3035fdd48fb1,  0 +   54bfb1,   exe, 6, [I8 ]
[ 6] 3035fdd48fe9,  0 +   54bfe9,   exe, 6, [I8 ]
[ 7] 3035fdd48ff1,  0 +   54bff1,   exe, 6, [I8 ]
[ 8] 3035fdd49029,  0 +   54c029,   exe, 6, [I8 ]
8> tag 0,1,2=speed_dirt_bike1
8> list
[ 0] 3035fdd48f29,  0 +   54bf29,   exe, 6, [I8 ],  speed_dirt_bike1
[ 1] 3035fdd48f31,  0 +   54bf31,   exe, 6, [I8 ],  speed_dirt_bike1
[ 2] 3035fdd48f69,  0 +   54bf69,   exe, 6, [I8 ],  speed_dirt_bike1
[ 3] 3035fdd48f71,  0 +   54bf71,   exe, 6, [I8 ]
[ 4] 3035fdd48fa9,  0 +   54bfa9,   exe, 6, [I8 ]
[ 5] 3035fdd48fb1,  0 +   54bfb1,   exe, 6, [I8 ]
[ 6] 3035fdd48fe9,  0 +   54bfe9,   exe, 6, [I8 ]
[ 7] 3035fdd48ff1,  0 +   54bff1,   exe, 6, [I8 ]
[ 8] 3035fdd49029,  0 +   54c029,   exe, 6, [I8 ]
8> tag 1,2,3=speed_p2_bike 
8> list
[ 0] 3035fdd48f29,  0 +   54bf29,   exe, 6, [I8 ],  speed_dirt_bike1
[ 1] 3035fdd48f31,  0 +   54bf31,   exe, 6, [I8 ],  speed_dirt_bike1,speed_p2_bike
[ 2] 3035fdd48f69,  0 +   54bf69,   exe, 6, [I8 ],  speed_dirt_bike1,speed_p2_bike
[ 3] 3035fdd48f71,  0 +   54bf71,   exe, 6, [I8 ],  speed_p2_bike
[ 4] 3035fdd48fa9,  0 +   54bfa9,   exe, 6, [I8 ]
[ 5] 3035fdd48fb1,  0 +   54bfb1,   exe, 6, [I8 ]
[ 6] 3035fdd48fe9,  0 +   54bfe9,   exe, 6, [I8 ]
[ 7] 3035fdd48ff1,  0 +   54bff1,   exe, 6, [I8 ]
[ 8] 3035fdd49029,  0 +   54c029,   exe, 6, [I8 ]
8> snapshot speed_*  player1Fast.snap
8>reset
....
1111>load snapshot player1Wepons.snap
1235> list wep*
[ 0] 3035fdd48f29,  0 +   54bf29,   exe, 6, [I8 ],  wepon_fireball
[ 1] 3035fdd48f31,  0 +   54bf31,   exe, 6, [I8 ],  wepon_spear
[ 2] 3035fdd48f69,  0 +   54bf69,   exe, 6, [I8 ],  wepon_dagger
[ 3] 3035fdd48f71,  0 +   54bf71,   exe, 6, [I8 ],  wepon_empty
1235>reset
.......
1111>load snapshot player1Fast.snap
1111>load snapshot player1Weapons.snap
1111>set speed*=8/10 **&** ## **<-- Notice the ampersand here**
set Interval id is 0
1111>stop interval 0
1111>set wepon_fireball=1 
1111>restart interval 0
1111>

## **snap file:**
3035fdd48f29,  0 +   54bf29,   exe, 6, [I8 ], defaultValue=, wepon_fireball
3035fdd48f31,  0 +   54bf31,   exe, 6, [I8 ],  defaultValue=, wepon_spear
3035fdd48f69,  0 +   54bf69,   exe, 6, [I8 ],  defaultValue=, wepon_dagger
3035fdd48f71,  0 +   54bf71,   exe, 6, [I8 ],  defaultValue=, wepon_empty
```

The ID column is removed as it is no longer relevant.
and defaultValue=  is an optional column that contains the value that would be set (if no value is specified at the command). So instead of 
`11>set speed*=8/10`
you could 
`11>set speed*=8/defaultValue`

I probably wouldn't actually call it "defaultValue".

Lastly I want an ability to script out the load, set, stop, restart commands in a function.
so I would type
1111> stopBikeFireAndRestart()

This would execute loading various snapshots setting memory regions, etc.. This script would need an ability to wait for a keystroke and probably sleep for some interval.

**Comment by sriemer** [comment]
We are in feature freeze right now as v0.16 is about to be released. Debian Stretch release schedule and therefore 2017-01-05 is the hard deadline.

How do you imagine to handle ASLR correctly with your snapshot storing and reloading?

We aren't in the old times of static unrandomized memory all put into the executable any more.

That feature worked with DOS, DOS games and an Action Replay module in the ISA slot the last time I remember.

Scanmem is no game trainer. Your tagging or rather "value name" is pointless for scanmem. Your value is likely at a different location any ways the next game run. A game trainer like ugtrain or if you only have static memory values in the executable without PIE (unlikely), then GameConqueror is the right place for names for values.

Scanmem can be scripted and you can use libscanmem for concurrent command execution. I don't see a real need for this in the interactive scanmem tool. Rather adding support for multi value locking/freezing would be nice e.g. by adding values to an internal locking/freezing list.

The regions can already be limited. I already did experiments with scans of parts of regions. In some special cases this might make sense, yes.

**Comment by asdf23** [comment]
> > 
> 
> We are in feature freeze right now as v0.16 is about to be released. Debian Stretch release schedule and therefore 2017-01-05 is the hard deadline.
> 
> How do you imagine to handle ASLR correctly with your snapshot storing and reloading?
> 
> We aren't in the old times of static unrandomized memory all put into the executable any more.
> 
> That feature worked with DOS, DOS games and an Action Replay module in the ISA slot the last time I remember.
> 
> Scanmem is no game trainer. Your tagging or rather "value name" is pointless for scanmem. Your value is likely at a different location any ways the next game run. A game trainer like ugtrain or if you only have static memory values in the executable without PIE (unlikely), then GameConqueror is the right place for names for values.
> 
> Scanmem can be scripted and you can use libscanmem for concurrent command execution. I don't see a real need for this in the interactive scanmem tool. Rather adding support for multi value locking/freezing would be nice e.g. by adding values to an internal locking/freezing list.
> 
> The regions can already be limited. I already did experiments with scans of parts of regions. In some special cases this might make sense, yes.
>

**Comment by sriemer** [comment]
Okay, so you're talking about the search snapshot. When loading it you can't be sure that values are still the same. So often you will find nothing in it or errors due to missing regions are caused. Recreation doesn't take that long or you forgot to exclude some too big regions. I don't get for what you would need this. Please provide a use case.

I was thinking about memory snapshot saving and reloading as a savegame feature for games which don't allow that instead. That can be useful but will be limited to the same game run and special conditions must be met like having at least the same space in all writable memory regions as we can only write to them but not extend or create them with scanmem. Needs to be evaluated in a POC.

The value names can be useful together with the multi value locking/freezing/watching. Concurrent execution might be useful to do memory search and value locking in parallel like GC does.

Scanmem reads commands from stdin.

```
echo -e "option scan_data_type string\n\"needle" | scanmem $(pidof mytool)
```
is an example for scanmem automation. Also multiple scanmem instances can be attached to the same process as scanmem handles attach errors gracefully and is only attached when needed.

**Comment by 12345ieee** [comment]
>> Scanmem can be scripted and you can use libscanmem for concurrent command execution.

> WooHoo!
> Can I use and do you have any examples ?

To have a well separated `libscanmem` you should probably wait for v0.16.
In the meantime, have a look at [backend.py](https://github.com/scanmem/scanmem/blob/master/gui/backend.py), which handles communication between `libscanmem` and `GameConqueror`, to get an idea of the API.

**Comment by 12345ieee** [comment]
Uhm, this is marked as v0.17. @asdf23 , what of this is still relevant/needed?
